repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java

```diff
@@ -1,96 +1,95 @@
-                ps.setString(1, tag.getName());
-                return ps;
-            }, keyHolder);
+    public List<Tag> findAll(int page, int size) {
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Tag> query = criteriaBuilder.createQuery(Tag.class);
+        Root<Tag> tagRoot = query.from(Tag.class);
 
-            tag.setId(Objects.requireNonNull(keyHolder.getKey()).longValue());
+        query.select(tagRoot);
 
-            return tag;
-        } catch (DataAccessException e) {
-            log.error("Error occurred while saving tag", e);
-            throw new TagOperationException("Error occurred while saving tag", e);
-        }
-    }
+        TypedQuery<Tag> typedQuery = entityManager.createQuery(query);
+        typedQuery.setFirstResult((page - 1) * size);
+        typedQuery.setMaxResults(size);
 
+        return typedQuery.getResultList();
+    }
 
     @Override
-    public Optional<Tag> findById(Long id) throws TagNotFoundException {
-        try {
-            Tag tag = jdbcTemplate.queryForObject(SELECT_BY_ID, new TagRowMapper(), id);
-            return Optional.ofNullable(tag);
-        } catch (EmptyResultDataAccessException e) {
-            log.error("Error occurred while getting tag with id = {}", id, e);
-            throw new TagNotFoundException("Tag not found with id: " + id, e);
-        }
+    public Optional<Tag> findById(Long id) {
+        return Optional.ofNullable(entityManager.find(Tag.class, id));
     }
 
     @Override
-    public Optional<Tag> findByName(String name) {
-        try {
-            Tag tag = jdbcTemplate.queryForObject(SELECT_BY_NAME, new TagRowMapper(), name);
-            return Optional.ofNullable(tag);
-        } catch (EmptyResultDataAccessException e) {
-            return Optional.empty();
-        }
+    public Tag save(Tag entity) {
+        entityManager.persist(entity);
+        return entity;
     }
 
-
     @Override
-    public List<Tag> findAll() throws TagNotFoundException {
-        try {
-            return jdbcTemplate.query(SELECT_ALL, new TagRowMapper());
-        } catch (DataAccessException e) {
-            log.error("Error occurred while getting all tags", e);
-            throw new TagNotFoundException("Error occurred while getting all tags", e);
-        }
+    public void delete(Tag tag) throws NotFoundException {
+        entityManager.remove(tag);
     }
 
     @Override
-    public void delete(Long id) throws TagOperationException {
-        try {
-            jdbcTemplate.update(DELETE, id);
-        } catch (DataAccessException e) {
-            log.error("Error occurred while deleting tag with id = {}", id, e);
-            throw new TagOperationException("Error occurred while deleting tag", e);
-        }
+    public List<Tag> findMostUsedTagOfUserWithHighestOrderCost(Long userId) {
+        String findMostUsedTagOfUserQuery = """
```

Comment:
Could you please create a new text file, add the query execution plan of this query and describe the steps?